README.md
# Solstack

**Solstack** is a library that enables the management of an application's control flow through a **state stack machine**.

A `Stack` `struct` holds any number of `State`s. When said `Stack` is ticked, it executes the main methods of the **topmost `State`**. This means that only the `State` at the top of the `Stack` is executed, while the ones below it are effectively **paused**. An exception are the `State`'s `on_shadow_tick` method introduced in `v0.3.0` which is always executed independently of the `State`'s position on the `Stack`.

- You can read [The Book](https://solmateus.gitbook.io/solstack/). 
- Take a look at the examples on the [git repository](https://github.com/solmateus/solstack). 
- Search documentation at the [crate's docs](https://crates.io/crates/solstack).
- Watch the project develop through its [changelog](https://github.com/solmateus/solstack/blob/main/CHANGELOG.md)!


> *Project is in early development, things may change around! Take a look at the changelog before updating. Some text may imply that things are done, when they're not, like The Book. There may be bugs, though the tests should guarantee there aren't many.*


## Overview

`State`s have all of the methods you'll need. Here's an overview:

* `on_start` — executed when this `State` is first pushed onto the `Stack`.
* `on_stop` — executed when this `State` is popped from the `Stack`.
* `on_pause` — executed when another `State` is pushed on top of this one in the `Stack`.
* `on_resume` — executed when this all of the `State`s above this one are popped from the `Stack`.
* `on_tick` — executed every time the `Stack` holding this `State` has it's `tick` method called.
* `on_shadow_tick` — same as `on_tick`, but is always run independently of this `State`'s position in the `Stack`.

To hop between `State`s you'll return a `Trans`ition `enum` from `on_tick` or similar methods, requesting the `Stack` to perform such transition next tick. Here's an overview:

* `Trans::None` — requests the `Stack` to do nothing.
* `Trans::Quit` — requests the `Stack` to `Pop` every `State` it is holding.
* `Trans::Push(Box::new(State))` — requests the `Stack` to `Push` the provided `State` on top.
* `Trans::Pop` — requests the `Stack` to pop the topmost `State` it is holding, deleting it.
* `Trans::Replace(Box::new(State))` — requests the `Stack` to `Pop` once and `Push` the provided `State`.
* `Trans::Isolate(Box::new(State))` — requests the `Stack` to `Pop` everything and `Push` the provided `State`.

> Transitions may be requested directly of the `Stack` or by returning a `Trans` from inside a `on_tick` or `on_shadow_tick` method of a `State`.


## Features

- An easy to implement `State` `trait`.
- An easy to use **State `Stack` Machine** `struct`.
- An easy to use set of **`Trans`itions between `States`**.

- New in `v0.3.0`:
  - `on_shadow_tick` is provided on `State`s; similar to `on_tick` but *is always executed independently of the `State`'s position* in the `Stack`.


## Use case

Imagine you're writing a game. You need a way of controlling the flow of your program. From the main menu to the game itself; from the game to the pause menu; or from the pause menu to quitting the program. 

***Solstack** will help you structure these `State`s and the  `Trans`itions between them*. 


Let's model something. Here are our `State`s (prefixed with `S`):

- `SMainMenu`: where the player lands on initializing the game.
- `SGame`: where the actual gameplay logic resides.
- `SPauseMenu`: where the player can save, resume playing or exit.


Since the `SMainMenu` is the first thing the user will encounter, we'll manually `Push` that `State` on our `Stack` when the program begins. Then we'll tick the `Stack` in a main loop until there are no `State`s inside it anymore.

> You can manually perform transitions on a `Stack` by using it's methods on your local instance.
>
> The loop can be achieved by using the `Stack`'s `is_running` method.

 The `Stack`, at the beginning of our program, will look like this: 

1.  `SMainMenu`


With only one sate on the `Stack`, it is on top; and so it will have it's methods called. `SMainMenu`'s logic is simple: when the player presses *START*, it requests the `Stack` to `Push` an `SGame`. If that happens, the `Stack` would look like this:

1. `SGame`
2. `SMainMenu`


Since only the topmost state is run by the stack's `tick`, `SMainMenu` just sits there. Now the player is enjoying their game; but they wish to pause! Well, inside `SGame` all we have to do is request the stack to `Push` an `SPauseMenu` if the player ever presses *ESC*. Simple! Let's see the `Stack` again:

1. `SPauseMenu`
2. `SGame`
3. `SMainMenu`


Now `SPauseMenu` is at the top. `SGame` will be **paused**; it's still there, but it is not being executed. Inside the `SPauseMenu` there should be logic saying that if the player presses *ESC* again, the `Stack` should `Pop`. `Pop`ping means removing or completely deleting the topmost `State` at the `Stack`. In this case, `SPauseMenu` itself. The `Stack` would then, again, look like this:

1. `SGame`
2. `SMainMenu`


Finally, `SGame` is at the top again! And so, it will resume **exactly** where it left off! If the player chooses to **quit the game**, you simply request the `Stack` to `Quit`, which will `Pop` every `State` it has, making the main loop end.


This concept can be extended to much bigger patterns; and hopefully you'll find joy in structuring your application with `solstack`!


## Get started

This is just a **very simplistic** example in code. Take a look at the links at the start of this page, specially the examples on the project's repository. The tests are also a good way of seeing how the library works internally.

```rust
use solstack::prelude::*;
use solstack::macros::*; // easy abstractions over boilerplate-y code.

// data available to `State`s for writing and reading
#[derive(Default)]
struct GameData {
    value: i32
}

// a `State` that does what it says
struct AddOneAndPrintState;
impl State<GameData> for AddOneAndPrintState {
    // run when this `State` is first pushed onto a `Stack`
    fn on_start(&mut self, data: &mut GameData) {
        data.value = 41;
        println!("on_start `make data be 41` > GameData({})", data.value);
    }

    // run every time the `Stack` is ticked.
    fn on_tick(&mut self, data: &mut GameData) -> Trans<GameData> {
        data.value += 1;
        println!("on_tick `add one to data` > GameData({})", data.value);
        Trans::None
    }
}

fn main() {
    // initializing
    let mut data = GameData::default();
    let mut stack = Stack::<GameData>::new();
    
    assert_eq!(data.value, 0);
    
    // manually pushing and ticking the `Stack`
    
    stack_push!(stack, data, AddOneAndPrintState);
    assert_eq!(data.value, 41);
    
    stack_tick!(stack, data);
    assert_eq!(data.value, 42);
    
    stack_tick!(stack, data);
    assert_eq!(data.value, 43);
    
}
```


## Thanks

> The documentation will always be up to date.

Thank you for using `solstack`!

*By Sol* <solmateusbraga@gmail.com>

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

*Performance notes are approximated results from criterion benchmarking.*


## [v.0.3.2] - 2023-03-28

### Improved

- **Performance increase** on previous version of roughly **30%**.
  - Refactor of `Stack::tick`; now uses iterators instead of for loops.

## [v0.3.1] - 2023-02-01

### Improved

- **Performance increase** on previous version of roughly **20%**.
  - Refactor of `Stack::tick`; now uses only one vector instead of two.

## [v0.3.0] - 2023-01-30

### Added

- Shadowing: methods on `State`s that are always executed on a `Stack`'s `tick`,
even if the `State` is not at the top.
  - `on_shadow_tick` for when the stack gets ticked through its `tick` method.

## [v0.2.0] - 2023-01-28

### Added

- `Trans`
  - `Replace`: pops and pushes a new state, effectively performing a 
  replacement.
  - `Isolate`: quits and pushes a new state, effectively isolating it as the 
  only one on the stack.
  - Macros and functions on `Stack` for the new transitions.

## [v0.1.0] - 2023-01-27

- First public release of `solstack`.

### Added

- Fully capable state stack machine.
- `State` callbacks added on this version:
  - `on_start` for when a state is pushed.
  - `on_stop` for when a state is popped.
  - `on_pause` for when a state has another pushed on top.
  - `on_resume` for when a stated gets resumed due to the one above being pushed.
  - `on_tick` for when the stack gets ticked through its `tick` method.
- `Trans`
  - `Push` that pushes a new state on top.
  - `Pop` that pops the topmost state out.
  - `None` that does not change the stack.
  - `Quit` that pops everything out.
- Examples
  - `guessing_game`
- Tests
  - `callback_counting`
  

tests/pop.rs
use solstack::macros::{stack_pop, trans_pop};
use solstack::{prelude::*, stack_push};

// No macros
struct SPop;
impl State<()> for SPop {
    fn on_tick(&mut self, _data: &mut ()) -> Trans<()> {
        Trans::Pop
    }
}

#[test]
fn popping() {
    let mut stack = Stack::new();

    stack.push(&mut (), Box::new(SPop));

    assert_eq!(stack.len(), 1);

    stack.pop(&mut ());

    assert_eq!(stack.len(), 0);
}

// With macros
struct SPopMacro;
impl State<()> for SPopMacro {
    fn on_tick(&mut self, _data: &mut ()) -> Trans<()> {
        trans_pop!()
    }
}

#[test]
fn popping_macro() {
    let mut stack = Stack::new();

    stack_push!(
        stack,
        (),
        SPopMacro,
        SPopMacro,
        SPopMacro,
        SPopMacro,
        SPopMacro,
        SPopMacro
    );

    assert_eq!(stack.len(), 6);

    stack_pop!(stack, ());

    assert_eq!(stack.len(), 5);

    stack_pop!(stack, (), 3);

    assert_eq!(stack.len(), 2);
}


tests/none.rs
use solstack::{prelude::*, stack_tick, stack_push};
use solstack::macros::trans_none;

// No macros
struct STest;
impl State<()> for STest {
    fn on_tick(&mut self, _data: &mut ()) -> Trans<()> {
        Trans::None
    }
}

#[test]
fn nonning() {
    let mut stack = Stack::new();

    stack.push(&mut (), Box::new(STest));
    stack.push(&mut (), Box::new(STest));
    stack.push(&mut (), Box::new(STest));

    stack.tick(&mut ());
    stack.tick(&mut ());
    stack.tick(&mut ());
        
    assert_eq!(stack.len(), 3);
}

// With macros
struct STestMacro;
impl State<()> for STestMacro {
    fn on_tick(&mut self, _data: &mut ()) -> Trans<()> {
        trans_none!()
    }
}

#[test]
fn nonning_macro() {
    let mut stack = Stack::new();

    stack_push!(stack, (), STestMacro);
    stack_push!(stack, (), STestMacro);
    stack_push!(stack, (), STestMacro);

    stack_tick!(stack, ());
    stack_tick!(stack, ());
    stack_tick!(stack, ());

    assert_eq!(stack.len(), 3);
}



tests/push.rs
use solstack::{prelude::*, stack_tick};
use solstack::macros::{stack_push, trans_push};

// No macros
struct SPush;
impl State<()> for SPush {
    fn on_tick(&mut self, _data: &mut ()) -> Trans<()> {
        Trans::Push(Box::new(SPush))
    }
}

#[test]
fn pushing() {
    let mut stack = Stack::new();

    stack.push(&mut (), Box::new(SPush));
    assert_eq!(stack.len(), 1);

    stack.tick(&mut ());
    assert_eq!(stack.len(), 2);

    stack.tick(&mut ());
    assert_eq!(stack.len(), 3);
}

// With macros
struct SPushMacro;
impl State<()> for SPushMacro {
    fn on_tick(&mut self, _data: &mut ()) -> Trans<()> {
        trans_push!(SPushMacro)
    }
}

#[test]
fn pushing_macro() {
    let mut stack = Stack::new();

    stack_push!(stack, (), SPushMacro);
    assert_eq!(stack.len(), 1);

    stack_tick!(stack, ());
    assert_eq!(stack.len(), 2);

    stack_tick!(stack, ());
    assert_eq!(stack.len(), 3);

    stack_push!(stack, (), SPushMacro, SPushMacro, SPushMacro);
    assert_eq!(stack.len(), 6);

    stack_tick!(stack, ());
    assert_eq!(stack.len(), 7);
}

src/main.rs
use solstack::prelude::*;

struct GameData {
    popped_dumbstate: bool,
    value: i32,
}

struct DumbState;
impl State<GameData> for DumbState {
    fn on_tick(
        &mut self,
        data: &mut GameData,
    ) -> solstack::trans::Trans<GameData> {
        data.value += 1;
        println!("Tick Popping DumbState {}", data.value);
        Trans::Pop
    }
}

struct GameState;
impl State<GameData> for GameState {
    fn on_start(&mut self, data: &mut GameData) {
        data.value += 1;
        println!("Start {}", data.value);
    }

    fn on_stop(&mut self, data: &mut GameData) {
        data.value += 1;
        println!("Stop {}", data.value);
    }

    fn on_pause(&mut self, data: &mut GameData) {
        data.value += 1;
        println!("Pause {}", data.value);
    }

    fn on_resume(&mut self, data: &mut GameData) {
        data.value += 1;
        println!("Resume {}", data.value);
    }

    fn on_tick(
        &mut self,
        data: &mut GameData,
    ) -> solstack::trans::Trans<GameData> {
        if data.popped_dumbstate {
            data.value += 1;
            println!("Tick Quitting {}", data.value);
            Trans::Quit
        } else {
            data.value += 1;
            data.popped_dumbstate = true;
            println!("Tick Pushing DumbState {}", data.value);
            Trans::Push(Box::new(DumbState))
        }
    }
}

fn main() {

    let mut data = GameData { value: 0, popped_dumbstate: false };
    let mut stack = Stack::<GameData>::new();

    stack.push(Box::new(GameState), &mut data);
    while stack.is_running() {
        stack.tick(&mut data);
    }

    assert_eq!(data.value, 7);

}


src/stack.rs
use crate::state::BoxState;
use crate::trans::Trans;

/// Handles the control flow of a stack of [`State`](crate::state::States).
/// Essentialy the `State Stack Machine`.
///
/// See the crates docs or the book for more information.
/// ### Generics
/// - D: Data available to all [`State`s](crate::state::State) to perform writes and reads to.
#[derive(Default)]
pub struct Stack<D> {
    stack: Vec<BoxState<D>>,
}

// Constructors
impl<D> Stack<D> {
    /// New empty stack state machine; a.k.a. [`Stack`].
    #[must_use]
    pub fn new() -> Self {
        Self { stack: vec![] }
    }
}

// Utilities
impl<D> Stack<D> {
    /// Checks if there is at least one state on the stack.
    #[must_use]
    pub fn is_running(&self) -> bool {
        !self.stack.is_empty()
    }

    /// Gets the length of the stack.
    #[must_use]
    pub fn len(&self) -> usize {
        self.stack.len()
    }

    /// Checks if the stack is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.stack.is_empty()
    }
}

// Pure transitions.
impl<D> Stack<D> {
    /// Pushes a new state above the current one.
    /// Effectively pauses the current state until everything above it is
    /// popped.
    pub fn push(&mut self, data: &mut D, mut state: BoxState<D>) {
        // Run current `State`'s `.on_pause()`.
        if let Some(curr_state) = self.stack.last_mut() {
            curr_state.on_pause(data);
        }

        // Starts the new state.
        state.on_start(data);
        // Pushes the new state onto the stack.
        self.stack.push(state);
    }

    /// Pops the current [`State`](crate::state::State) from the stack.
    pub fn pop(&mut self, data: &mut D) {
        // Runs this `State`'s `.on_stop()`.
        // Also pops it from the stack (rhs of if-let).
        if let Some(mut curr_state) = self.stack.pop() {
            curr_state.on_stop(data);
        }

        // Runs the `State` below's `.on_resume()`.
        if let Some(state) = self.stack.last_mut() {
            state.on_resume(data);
        }
    }

    /// Pops everything from the stack.
    /// Effectively ends the state stack machine.
    pub fn quit(&mut self, data: &mut D) {
        // Pops everything out of the stack.
        // Also calls `.on_stop()` for all of the stack (rhs of while-let).
        while let Some(mut state) = self.stack.pop() {
            state.on_stop(data);
        }
    }
}

// Composed transitions.
impl<D> Stack<D> {
    /// Pops and pushes a new state.
    /// Effectively replaces the current state with a new one.
    #[inline]
    pub fn replace(&mut self, data: &mut D, state: BoxState<D>) {
        self.pop(data);
        self.push(data, state);
    }

    /// Pops every state from the stack and pushes a new one.
    /// Effectively isolates it as the only state on the stack.
    #[inline]
    pub fn isolate(&mut self, data: &mut D, state: BoxState<D>) {
        self.quit(data);
        self.push(data, state);
    }
}

impl<D> Stack<D> {
    #[inline]
    fn transition(&mut self, data: &mut D, trans: Trans<D>) {
        match trans {
            Trans::Push(state) => self.push(data, state),
            Trans::Pop => self.pop(data),
            Trans::Replace(state) => self.replace(data, state),
            Trans::Isolate(state) => self.isolate(data, state),
            Trans::None => (),
            Trans::Quit => self.quit(data),
        };
    }

    /// Ticks the current (topmost at the stack) state once.
    pub fn tick(&mut self, data: &mut D) {
        // Initializing an empty vector of transitions.
        let mut transitions: Vec<Trans<D>> = vec![];

        // Looping through every state.
        // Performing `on_shadow_tick` on every state.
        // Storing the transitions returned on the `transitions` vector.
        self.stack.iter_mut().for_each(|state| {
            transitions.push(state.on_shadow_tick(data))
        });

        // Getting the topmost state at the sack.
        // Ticking that state.
        // Storing the transition returned on the `transitions` vector.
        if let Some(state) = self.stack.last_mut() {
            transitions.push(state.on_tick(data));
        }

        // Performing the transitions stored in the `transitions` vector.
        transitions.into_iter().for_each(|trans| {
            self.transition(data, trans);
        });
    }
}



src/state.rs
use crate::trans::Trans;

#[allow(clippy::module_name_repetitions)]
pub type BoxState<D> = Box<dyn State<D>>;

/// Describes an application state.
///
/// See the crates docs or the book for more information.
/// ### Generics
/// - D: Data available to all [`State`](crate::state::State)s to perform actions to.
pub trait State<D> {

    /// Called when this [`State`](crate::state::State) first enters the stack.
    fn on_start(&mut self, _data: &mut D) {}

    /// Called when this [`State`](crate::state::State) is popped from the stack.
    fn on_stop(&mut self, _data: &mut D) {}

    /// Called when this [`State`](crate::state::State) loses its topmost position in the stack to
    /// another [`State`](crate::state::State) that has been pushed ontop of it.
    fn on_pause(&mut self, _data: &mut D) {}

    /// Called when this [`State`](crate::state::State) regains the first position in the stack.
    /// Id est, when other [`State`](crate::state::State)s above it are popped.
    fn on_resume(&mut self, _data: &mut D) {}

    /// Represents a single tick/update.
    /// It's called when the [`Stack`](crate::stack::Stack)'s `.tick()` is called.
    /// Your loop logic should call your stack's tick, not this directly.
    fn on_tick(&mut self, _data: &mut D) -> Trans<D> { 

        Trans::None
    }

    /// Represents a single tick/update independent of this [`State`](crate::state::State) 
    /// position in the stack.
    /// It's called when the [`Stack`](crate::stack::Stack)'s `.tick()` is called.
    /// NOTE: `on_shadow_tick` is called BEFORE `on_tick`.
    fn on_shadow_tick(&mut self, _data: &mut D) -> Trans<D> {

        Trans::None
    }
}



src/trans.rs
use crate::state::BoxState;

/// A transition between [`State`](crate::state::State)s.
///
/// See the crates docs or the book for more information.
/// ### Generics
/// - D: Data available to all [`State`](crate::state::State)s to perform actions to.
pub enum Trans<D> {
    /// Pushes a new state above the current one.
    /// Effectively pauses the current state until everything above it is
    /// popped.
    Push(BoxState<D>),

    /// Pops the current state from the stack.
    Pop,

    /// Pops and pushes a new state.
    /// Effectively replaces the current state with a new one.
    Replace(BoxState<D>),

    /// Pops every state from the stack and pushes a new one.
    /// Effectively isolates it as the only state on the stack.
    Isolate(BoxState<D>),

    /// Pops everything from the stack.
    /// Effectively ends the state stack machine.
    Quit,

    /// Does nothing to the stack.
    /// Effectively keeps the current state and the stack the way it is.
    None,
}


src/macros.rs
// Stack

/// Alternative to [`stack.push(&mut data, Box::new(FooState {}))`](crate::stack::Stack::push).
/// Use `stack_push!(stack, data, FooState {}, BarState {}, ...)`
/// - Accepts multiple states. In which case it will push all of the provided.
/// OBS.: Don't use `&mut data` as a parameter, but simply `data`.
#[macro_export]
macro_rules! stack_push {
    ($stack:expr, $data:expr, $($state:expr),+) => {
        $($stack.push(&mut $data, Box::new($state));)+
    }
}

pub use stack_push;

/// Alternative to [`stack.pop(&mut data)`](crate::stack::Stack::pop).
/// Use `stack_pop!(stack, data)`
/// Or `stack_pop!(stack, data, 3)` 3 being the amount of states to pop.
/// OBS.: Don't use `&mut data` as a parameter, but simply `data`.
/// - Accepts an `amount` of times to pop.
#[macro_export]
macro_rules! stack_pop {
    ($stack:expr, $data:expr) => {
        $stack.pop(&mut $data)
    };

    ($stack:expr, $data:expr, $amount:expr) => {
        for _ in 0..$amount {
            $stack.pop(&mut $data)
        }
    };
}

pub use stack_pop;

/// Alternative to [`stack.replace(&mut data, Box::new(FooState {}))`](crate::stack::Stack::replace).
/// Use `stack_replace!(stack, data, FooState {}, BarState {}, ...)`
/// OBS.: Don't use `&mut data` as a parameter, but simply `data`.
/// - Accepts multiple `State`s. In which case it will pop once and push the
/// provided states.
#[macro_export]
macro_rules! stack_replace {
    ($stack:expr, $data:expr, $($state:expr),+) => {
        $(
            $stack.pop(&mut $data);
            $stack.push(&mut $data, Box::new($state));
        )+
    }
}

pub use stack_replace;

/// Alternative to [`stack.isolate(&mut data, Box::new(FooState {}))`](crate::stack::Stack::isolate).
/// Use `stack_isolate!(stack, data, FooState {}, BarState {}, ...)`
/// OBS.: Don't use `&mut data` as a parameter, but simply `data`.
/// - Accepts multiple `State`s. In which case it will pop everything from the 
/// stack and then push the provided states.
#[macro_export]
macro_rules! stack_isolate{
    ($stack:expr, $data:expr, $($state:expr),+) => {
        // Pops everything from the stack.
        $stack.quit(&mut $data);
        // Adds the provided states for isolation.
        $($stack.push(&mut $data, Box::new($state));)+
    }
}

pub use stack_isolate;

/// Alternative to [`stack.quit(&mut data)`](crate::stack::Stack::quit).
/// Use `stack_quit!(stack, data).`
/// OBS.: Don't use `&mut data` as a parameter, but simply `data`.
#[macro_export]
macro_rules! stack_quit {
    ($stack:expr, $data:expr) => {
        $stack.quit(&mut $data)
    };
}

pub use stack_quit;

/// Alternative to [`stack.tick(&mut data)`](crate::stack::Stack::tick).
/// Use `stack_tick!(stack, data)`.
/// OBS.: Don't use `&mut data` as a parameter, but simply `data`.
#[macro_export]
macro_rules! stack_tick {
    ($stack:expr, $data:expr) => {
        $stack.tick(&mut $data)
    };
}

pub use stack_tick;

// Transitions

/// Alternative to [`Trans::Push(Box::new(FooState {}))`](crate::trans::Trans::Push).
/// Use `trans_push!(FooState {})`
#[allow(clippy::crate_in_macro_def)]
#[macro_export]
macro_rules! trans_push {
    ($state:expr) => {
        solstack::trans::Trans::Push(Box::new($state))
    };
}

pub use trans_push;

/// Alternative to [`Trans::Pop`](crate::trans::Trans::Pop).
/// Use `trans_pop!()`
#[allow(clippy::crate_in_macro_def)]
#[macro_export]
macro_rules! trans_pop {
    () => {
        solstack::trans::Trans::Pop
    };
}

pub use trans_pop;

/// Alternative to [`Trans::Replace(Box::new(FooState {}))`](crate::trans::Trans::Replace).
/// Use `trans_replace!(FooState {})`
#[allow(clippy::crate_in_macro_def)]
#[macro_export]
macro_rules! trans_replace{
    ($state:expr) => {
        solstack::trans::Trans::Replace(Box::new($state))
    };
}

pub use trans_replace;

/// Alternative to [`Trans::Isolate(Box::new(FooState {}))`](crate::trans::Trans::Isolate).
/// Use `trans_isolate!(FooState {})`
#[allow(clippy::crate_in_macro_def)]
#[macro_export]
macro_rules! trans_isolate{
    ($state:expr) => {
        solstack::trans::Trans::Isolate(Box::new($state))
    };
}

pub use trans_isolate;

/// Alternative to [`Trans::None`](crate::trans::Trans::None).
/// Use `trans_none!()`
#[allow(clippy::crate_in_macro_def)]
#[macro_export]
macro_rules! trans_none {
    () => {
        solstack::trans::Trans::None
    };
}

pub use trans_none;

/// Alternative to [`Trans::Quit`](crate::trans::Trans::Quit).
/// Use `trans_quit!()`
#[allow(clippy::crate_in_macro_def)]
#[macro_export]
macro_rules! trans_quit {
    () => {
        solstack::trans::Trans::Quit
    };
}

pub use trans_quit;


benches/ticking.rs
use criterion::{criterion_group, criterion_main, Criterion};

use solstack::{prelude::*, stack_push};

struct MakeItA;
struct MakeItB;
struct Dummy;

impl State<String> for MakeItA {
    fn on_tick(&mut self, data: &mut String) -> Trans<String> {
        *data = "A".to_owned();
        Trans::Replace(Box::new(MakeItB))
    }
}

impl State<String> for MakeItB {
    fn on_tick(&mut self, data: &mut String) -> Trans<String> {
        *data = "B".to_owned();
        Trans::Replace(Box::new(MakeItA))
    }
}

impl State<String> for Dummy {}

pub fn tick(c: &mut Criterion) {
    let mut data = String::new();
    let mut stack = Stack::<String>::new();

    c.bench_function("tick", |b| {
        b.iter(|| {
            stack_push!(stack, data, Dummy);
            stack.tick(&mut data)
        })
    });
}

criterion_group!(benches, tick);
criterion_main!(benches);


src/lib.rs
//! # Solstack
//!
//! **Solstack** is a library that enables the management of an application's control flow through a **state stack machine**.
//!
//! A `Stack` `struct` holds any number of `State`s. When said `Stack` is ticked, it executes the main methods of the **topmost `State`**. This means that only the `State` at the top of the `Stack` is executed, while the ones below it are effectively **paused**. An exception are the `State`'s `on_shadow_tick` method introduced in `v0.3.0` which is always executed independently of the `State`'s position on the `Stack`.
//!
//! - You can read [The Book](https://solmateus.gitbook.io/solstack/). 
//! - Take a look at the examples on the [git repository](https://github.com/solmateus/solstack). 
//! - Search documentation at the [crate's docs](https://crates.io/crates/solstack).
//! - Watch the project develop through its [changelog](https://github.com/solmateus/solstack/blob/main/CHANGELOG.md)!
//!
//!
//! > *Project is in early development, things may change around! Take a look at the changelog before updating. Some text may imply that things are done, when they're not, like The Book. There may be bugs, though the tests should guarantee there aren't many.*
//!
//!
//! ## Overview
//!
//! `State`s have all of the methods you'll need. Here's an overview:
//!
//! * `on_start` — executed when this `State` is first pushed onto the `Stack`.
//! * `on_stop` — executed when this `State` is popped from the `Stack`.
//! * `on_pause` — executed when another `State` is pushed on top of this one in the `Stack`.
//! * `on_resume` — executed when this all of the `State`s above this one are popped from the `Stack`.
//! * `on_tick` — executed every time the `Stack` holding this `State` has it's `tick` method called.
//! * `on_shadow_tick` — same as `on_tick`, but is always run independently of this `State`'s position in the `Stack`.
//!
//! To hop between `State`s you'll return a `Trans`ition `enum` from `on_tick` or similar methods, requesting the `Stack` to perform such transition next tick. Here's an overview:
//!
//! * `Trans::None` — requests the `Stack` to do nothing.
//! * `Trans::Quit` — requests the `Stack` to `Pop` every `State` it is holding.
//! * `Trans::Push(Box::new(State))` — requests the `Stack` to `Push` the provided `State` on top.
//! * `Trans::Pop` — requests the `Stack` to pop the topmost `State` it is holding, deleting it.
//! * `Trans::Replace(Box::new(State))` — requests the `Stack` to `Pop` once and `Push` the provided `State`.
//! * `Trans::Isolate(Box::new(State))` — requests the `Stack` to `Pop` everything and `Push` the provided `State`.
//!
//! > Transitions may be requested directly of the `Stack` or by returning a `Trans` from inside a `on_tick` or `on_shadow_tick` method of a `State`.
//!
//!
//! ## Features
//!
//! - An easy to implement `State` `trait`.
//! - An easy to use **State `Stack` Machine** `struct`.
//! - An easy to use set of **`Trans`itions between `States`**.
//!
//! - New in `v0.3.0`:
//!   - `on_shadow_tick` is provided on `State`s; similar to `on_tick` but *is always executed independently of the `State`'s position* in the `Stack`.
//!
//!
//! ## Use case
//!
//! Imagine you're writing a game. You need a way of controlling the flow of your program. From the main menu to the game itself; from the game to the pause menu; or from the pause menu to quitting the program. 
//!
//! ***Solstack** will help you structure these `State`s and the  `Trans`itions between them*. 
//!
//!
//! Let's model something. Here are our `State`s (prefixed with `S`):
//!
//! - `SMainMenu`: where the player lands on initializing the game.
//! - `SGame`: where the actual gameplay logic resides.
//! - `SPauseMenu`: where the player can save, resume playing or exit.
//!
//!
//! Since the `SMainMenu` is the first thing the user will encounter, we'll manually `Push` that `State` on our `Stack` when the program begins. Then we'll tick the `Stack` in a main loop until there are no `State`s inside it anymore.
//!
//! > You can manually perform transitions on a `Stack` by using it's methods on your local instance.
//! >
//! > The loop can be achieved by using the `Stack`'s `is_running` method.
//!
//!  The `Stack`, at the beginning of our program, will look like this: 
//!
//! 1.  `SMainMenu`
//!
//!
//! With only one sate on the `Stack`, it is on top; and so it will have it's methods called. `SMainMenu`'s logic is simple: when the player presses *START*, it requests the `Stack` to `Push` an `SGame`. If that happens, the `Stack` would look like this:
//!
//! 1. `SGame`
//! 2. `SMainMenu`
//!
//!
//! Since only the topmost state is run by the stack's `tick`, `SMainMenu` just sits there. Now the player is enjoying their game; but they wish to pause! Well, inside `SGame` all we have to do is request the stack to `Push` an `SPauseMenu` if the player ever presses *ESC*. Simple! Let's see the `Stack` again:
//!
//! 1. `SPauseMenu`
//! 2. `SGame`
//! 3. `SMainMenu`
//!
//!
//! Now `SPauseMenu` is at the top. `SGame` will be **paused**; it's still there, but it is not being executed. Inside the `SPauseMenu` there should be logic saying that if the player presses *ESC* again, the `Stack` should `Pop`. `Pop`ping means removing or completely deleting the topmost `State` at the `Stack`. In this case, `SPauseMenu` itself. The `Stack` would then, again, look like this:
//!
//! 1. `SGame`
//! 2. `SMainMenu`
//!
//!
//! Finally, `SGame` is at the top again! And so, it will resume **exactly** where it left off! If the player chooses to **quit the game**, you simply request the `Stack` to `Quit`, which will `Pop` every `State` it has, making the main loop end.
//!
//!
//! This concept can be extended to much bigger patterns; and hopefully you'll find joy in structuring your application with `solstack`!
//!
//!
//! ## Get started
//!
//! This is just a **very simplistic** example in code. Take a look at the links at the start of this page, specially the examples on the project's repository. The tests are also a good way of seeing how the library works internally.
//!
//! ```rust
//! use solstack::prelude::*;
//! use solstack::macros::*; // easy abstractions over boilerplate-y code.
//!
//! // data available to `State`s for writing and reading
//! #[derive(Default)]
//! struct GameData {
//!     value: i32
//! }
//!
//! // a `State` that does what it says
//! struct AddOneAndPrintState;
//! impl State<GameData> for AddOneAndPrintState {
//!     // run when this `State` is first pushed onto a `Stack`
//!     fn on_start(&mut self, data: &mut GameData) {
//!         data.value = 41;
//!         println!("on_start `make data be 41` > GameData({})", data.value);
//!     }
//!
//!     // run every time the `Stack` is ticked.
//!     fn on_tick(&mut self, data: &mut GameData) -> Trans<GameData> {
//!         data.value += 1;
//!         println!("on_tick `add one to data` > GameData({})", data.value);
//!         Trans::None
//!     }
//! }
//!
//! fn main() {
//!     // initializing
//!     let mut data = GameData::default();
//!     let mut stack = Stack::<GameData>::new();
//!    
//!     assert_eq!(data.value, 0);
//!    
//!     // manually pushing and ticking the `Stack`
//!    
//!     stack_push!(stack, data, AddOneAndPrintState);
//!     assert_eq!(data.value, 41);
//!    
//!     stack_tick!(stack, data);
//!     assert_eq!(data.value, 42);
//!    
//!     stack_tick!(stack, data);
//!     assert_eq!(data.value, 43);
//!    
//! }
//! ```
//!
//!
//! ## Thanks
//!
//! > The documentation will always be up to date.
//!
//! Thank you for using `solstack`!
//!
//! *By Sol* <solmateusbraga@gmail.com>

#![warn(clippy::pedantic)]
#![allow(clippy::needless_doctest_main)]

pub mod macros;
pub mod stack;
pub mod trans;
pub mod state;

pub mod prelude {

    pub use crate::stack::Stack;
    pub use crate::trans::Trans;
    pub use crate::state::State;
}
